### ISSUE TITLE
Pod-to-pod networking broken

### ISSUE THREAD

**Issue by MikaelCluseau** [BenTheElder/kubernetes#3]
From my tests, pod-to-pod networking through service IPs doesn't work. I'm pretty sure it's because the packet are not SNATed:

```
$ time curl -i mail.master
HTTP/1.1 404 Not Found
Content-Type: application/octet-stream
Content-Length: 58
Server: Jetty(9.2.9.v20150224)

Could not find resource for full path: http://mail.master/
real	0m0.019s
user	0m0.008s
sys	0m0.000s
$ kubectl --namespace=master exec keystone-zdr02 -- wget mail
--2015-07-11 00:08:44--  http://mail/
Resolving mail (mail)... 10.107.140.79
Connecting to mail (mail)|10.107.140.79|:80... failed: Connection timed out.
Retrying.
```

I'll try to add a MASQERADE rule. I know this will hide the real source IP from the pod, but it's not worse than with the original proxy ;)

**Comment by MikaelCluseau** [comment]
("opened" in #2)

I used MASQUERADE because otherwise the source IP is the real one, thus the destination pod replies with its own IP instead of the service IP, which causes the packet to be rejected.

**Comment by MikaelCluseau** [comment]
IMHO, what we need is a way to have the reply packet catched by the kernel's connection tracking, and apply a rule to set the service IP as the source.

I used MASQUERADE to do that as a side effect, since it replaces the source IP with the outgoing interface IP, which make the target pod to reply to the host, which then catches it with the conntrack, and replaces the reply's source IP to the service IP.

There may be a way (with ebtables?) to catch the packet as it goes through the pod's bridge (`flannel.1` for me). This would make the target pod see "source pod IP -> target pod IP" and the source pod see "source pod IP -> service IP". The problem is... I don't know enough of networking to do that.

**Comment by BenTheElder** [comment]
That makes sense.

I don't know how to do the last part either but I'll look into it as that sounds like what we would want ideally, but `MASQUERADE` seems like a sane solution in the meantime.

**Comment by MikaelCluseau** [comment]
`MASQUERADE`could be limited to traffic going from the pod to the service IPs, that would allow us to solve the problem with only one rule like `-j MASQUERADE -s pod_network -d service_network`.

**Comment by MikaelCluseau** [comment]
I found this, I'm having a look: http://ebtables.netfilter.org/br_fw_ia/br_fw_ia.html

**Comment by BenTheElder** [comment]
I think the limited `MASQUERADE` rule seems like the best solution for now.

**Comment by MikaelCluseau** [comment]
Yes, I'm trying to implement it on my "development production" single-node cluster.

**Comment by BenTheElder** [comment]
Thank you. I'm still reading through the options for manipulating Source Addresses but so far that sounds ideal.

**Comment by MikaelCluseau** [comment]
The `MASQUERADE` rule I gave is not ok, POSTROUTING gets the pods IPs, so it works with this ugly thing:

    iptables -t nat -I POSTROUTING -j MASQUERADE -s 10.244.0.0/16 -d 10.244.0.0/16

I may use the recent pattern finally ;)

**Comment by MikaelCluseau** [comment]
The following additionnal rules make it work for me:

```
iptables -t nat -I PREROUTING -s pod_network -d service_network -m state --state NEW -m recent --name KUBE-MASQUERADE --set
iptables -t nat -I POSTROUTING -m recent --name KUBE-MASQUERADE --rcheck --seconds 1 --reap -j MASQUERADE
```

**Comment by BenTheElder** [comment]
Great!

I guess we'll have to move those out to more chains so that we can update these in `syncProxyRules()` and set the chains up in `iptablesinit`.

As I read the snippet you posted, `service_network` is the target ip's (?); then is the `pod_network` then the proxyIP (host/listening ip) on each node or?

**Comment by MikaelCluseau** [comment]
Yes, I think at least one more chain is required since no kubernetes' specific chain is targeted from `POSTROUTING`.

The service network is the `--service-cluster-ip-range` of apiserver. The pod network is provided by flannel in my setup; it's the `--bip` of the docker daemon. I don't how you can get this information though... an alternative way of achieve the same result would be to use an ipset or to put as many rules as there are pods on the host.

**Comment by BenTheElder** [comment]
Hmm, ok well we'll have to figure that out then. Thanks again. I'll look into it more tomorrow.

**Comment by BenTheElder** [comment]
@Symmetric, any thoughts on this? I know you were looking forward to the source ip preservation but this seems necessary, we will probably need a `MASQUERADE` rule.

Discussion started here: https://github.com/BenTheElder/kubernetes/pull/2#issuecomment-120555518

**Comment by Symmetric** [comment]
This is puzzling, I'll have to dig deeper here -- when I tested Tim's basic rules, and a previous iteration of the PR, things seemed to work fine. My understanding of DNAT is that the return packet should be un-DNAT'd so the src and dst IPs at the originator will be correct (i.e. the response packet should have the Service IP as the source once iptables is finished with it).

Have you reproduced this in a vanilla Kubernetes environment? Sounds like Mikeal is using Flannel, which might be complicating things.

I should have time to dig in a day or two. But yes, it would be desirable to keep the true src IP of service packets, since that allows us to do proper firewalling at the destination.

**Comment by BenTheElder** [comment]
I have not, so far in my own testing all of the problems i've found appear to have been related to vagrant not the proxy. I'm experimenting with the current source now, but as of yet have not reproduced the issue.

I'm not very familiar with Flannel, perhaps it is an issue specific to using Flannel.

**Comment by BenTheElder** [comment]
Latest full e2e results:
```
Summarizing 3 Failures:

[Fail] Kubectl client Simple pod [It] should support port-forward 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/kubectl.go:197

[Fail] Services [It] should be able to create a functioning NodePort service 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:1036

[Fail] ServiceAccounts [It] should mount an API token into pods 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:915

Ran 51 of 98 Specs in 2844.044 seconds
FAIL! -- 48 Passed | 3 Failed | 2 Pending | 45 Skipped --- FAIL: TestE2E (2849.25s)
```

The first and last failures appear to be unrelated to the proxy, the second one however...

```
• Failure [330.666 seconds]
Services
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:798
  should be able to create a functioning NodePort service [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:327

  Error waiting for the url http://10.245.1.3:31673 to be reachable
  Expected error:
      <*errors.errorString | 0x209345230>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred

  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:1036
```

I will switch to master and compare results.

It should be noted that a number of tests are now skipped on vagrant because they don't make sense, so "full" is actually missing a few tests that would be run on e.g. GCE.

**Comment by BenTheElder** [comment]
For reference, the currently comparable master gives only one failure (the first).
```
Summarizing 1 Failure:

[Fail] Kubectl client Simple pod [It] should support port-forward 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/kubectl.go:197

Ran 51 of 98 Specs in 2225.808 seconds
FAIL! -- 50 Passed | 1 Failed | 2 Pending | 45 Skipped --- FAIL: TestE2E (2231.01s)
```

**Comment by MikaelCluseau** [comment]
@Symmetric I think it is not going through the conntrack code because it's at IP level, and replying on a local bridge doesn't go through IP code because it doesn't have to be routed. See [here](http://ebtables.netfilter.org/br_fw_ia/br_fw_ia.html#section2):

> The bridge's decision for a frame (as seen on Figure 2b) can be one of these:
>* bridge it, if the destination MAC address is on another side of the bridge;
>* flood it over all the forwarding bridge ports, if the position of the box with the destination MAC is unknown to the bridge;
>* pass it to the higher protocol code (the IP code), if the destination MAC address is that of the bridge or of one of its ports;
>* ignore it, if the destination MAC address is located on the same side of the bridge.

So, something may be possible at the bridge (ebtables) level, or somewhere unknown to me. I will try to upgrade my flannel, just in case, but as per the 3rd rule quoted here, they would need to do something very special. The standard configuration as per kubernetes getting-started guide for CoreOS creates a vxlan to allow containers to communicate on a multi-host bridge.

**Comment by MikaelCluseau** [comment]
I completely stopped and removed flannel (including the vxlan interface it created). Masquerading is still required for pod-to-pod networking to work.

I also tried to set the reply's MAC address to the docker bridge's IP using DNAT/SNAT rules at the ebtables level without success.

For SNAT I used:
```
ebtables -t broute -A BROUTING --ip-destination 10.107.0.0/24 -p 0x0800 -j mark --mark-set 1
ebtables -t nat -A POSTROUTING --mark 1 -j snat --to-source 02:0e:e4:cc:0b:cf
```

For DNAT I used:
```
ebtables -t nat -A PREROUTING --in-interface docker0 --ip-destination 10.107.0.0/16 -p 0x0800 -j redirect
```

Having any of these, or both, didn't solve the problem :(

**Comment by Symmetric** [comment]
Finally got a chance to look at this. As stated above, the problem is that traffic on the bridge isn't going through iptables. This is configurable behaviour (see [bridge-nf](http://ebtables.netfilter.org/misc/brnf-faq.html) for details).

I set this on the node:

```
sudo  sysctl -w net.bridge.bridge-nf-call-iptables=1
sudo  sysctl -w net.bridge.bridge-nf-call-ip6tables=1
```

And the rules work again. @MikaelCluseau, can you give this a try and see if it fixes things in your environment?

bridge-nf seems to have been introduced in kernel 2.6.1, which is satisfied by the k8s requirement for RHEL 2.6.32-431/other 3.10, so I think it's safe to assume this flag is present.

**Comment by BenTheElder** [comment]
Thanks for looking into it @Symmetric.

I'm in the process of updating to the latest master and re-running everything and looking at those e2e tests, since I am still seeing failure on proxy related (NodePort) tests which suggests that the rules may not be fully functioning as kubernetes expects still. I'll check in with findings later.

Hopefully it's just a flake and the configuration change fixes @MikaelCluseau's issues.

**Comment by MikaelCluseau** [comment]
@Symmetric ok, that was somewhere unknown to me :-) Under CoreOS, the nf-bridge module isn't loaded by default, and loading it is enough to have the sysctls set to 1:
```
# modprobe br-netfilter
# cat /proc/sys/net/bridge/bridge-nf-call-iptables 
1
```
After that, you are right, the bridge traffic goes through the IP stack and pod-to-pod networking works as expected, this is really great, thanks! :-)

The getting-started for CoreOS will have to be updated so that node.yaml has the following added:
```
write_files:
  - path: /etc/modules-load.d/kubernetes.conf
    content: br_netfilter
coreos:
  units:
    - name: systemd-modules-load.service
      command: restart
```

For reference, the output of my test:
```
$ kubectl --namespace=master exec mail-9j4y6 -- curl --connect-timeout0 keytone:35357 >/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   765  100   765    0     0  88357      0 --:--:-- --:--:-- --:--:--  106k
$ kubectl --namespace=master logs keystone-ea7b4 |tail -n1
2015-07-15 19:07:02.884 20 INFO eventlet.wsgi.server [-] 10.244.19.78 - - [15/Jul/2015 19:07:02] "GET / HTTP/1.1" 300 932 0.001088
```

**Comment by MikaelCluseau** [comment]
This is not CoreOS but Linux >=3.18 specific, as can be seen [here](http://ebtables.netfilter.org/documentation/bridge-nf.html):
> Since Linux kernel 3.18-rc1, you have to modprobe br_netfilter to enable bridge-netfilter.

**Comment by Symmetric** [comment]
@MikaelCluseau great news, glad we didn't have to get into the gnarly iptables rules!

@BenTheElder, I think we need to make sure that this flag is set on all kubernetes clusters when upgrading to the version that includes this patch. Not sure how to go about that; are you aware of a mechanism in kubelet to set sysctls? Or do we need to set this in the salt files (and every other getting started guide)? Obviously the former is vastly preferable.

There's also a question about how to handle upgrades; before upgrading an existing k8s cluster to the version that introduces this patch, operators will need to make sure the module is loaded and the sysctls set. Perhaps one to follow up with the k8s guys, as I don't know if there is a precedent for this issue that we can copy.

**Comment by BenTheElder** [comment]
@MikaelCluseau excellent!

@Symmetric I'm not aware of one, as far as I can tell all of the that is done in the salt files. A quick search suggests that `sysctl` does not exist in any of the go source but some is in the guides and salt files.
I think we may have to open an issue upstream with the k8s guys.

**Comment by MikaelCluseau** [comment]
Huho, I have last thing that will very probably require MASQUERADING, for a very specific case: if a pod connects to a service that it serves, it will its own IP from the source side, will answer to that directly, and is not catched by connection tracking. We have that on a replicated service that tries to check if it is a duplicate by connecting to its advertised host name, which is bound to the service IP: the connection stay in SYN_SENT state. netstat in the pod gives:

```
Proto Recv-Q Send-Q Local Address           Foreign Address         State
[...]       
tcp6       0      1 10.244.19.74:49381      10.107.162.84:9092      SYN_SENT
```

10.107.162.84 is the service IP bound to 10.244.19.74.

This will require a masquerading rule in the form iptables -t nat -I POSTROUTING -s <pod IP> -d <pod IP> -j MASQUERADE or something like that. I still have to test that though, but I wanted to share if anyone here have an idea like "hey, you need the br_netfilter module" :)

**Comment by BenTheElder** [comment]
Hmm. Perhaps the rule can be added separately/externally to the proxy? That would still be less than ideal though. I'm not sure of a more elegant solution but it would be nice to avoid falsifying the source address in the general case.

**Comment by MikaelCluseau** [comment]
Erf... again I used &lt;&gt; without previewing... but anyway this rules doesn't seems to work. I'm still looking for the best solution. The rule was intented exactly for the case `--source pod_IP --destination pod_IP`, so very specific (will not falsify anything but podA to podA traffic).

**Comment by BenTheElder** [comment]
Ah, yes that does seem like a very reasonable rule. @Symmetric would this be problematic for calico?

**Comment by MikaelCluseau** [comment]
I tried many ways... this kind of rule doesn't work. From what I could understand using `-j TRACE` on the connection packet, the packet stops traversing the rules after the matching `DNAT` rule, because it sets a conntrack entry. Which mean the `MASQUERADE` rules are not applied after the DNAT, and if we don't DNAT we don't set the right target for the connection.

The only solution I could find was to fallback to some userland proxy and a REDIRECT rule, which is a quite disappointing fallback :( I continue to look for a better solution.

**Comment by BenTheElder** [comment]
Hmm, sorry about that. I'm not really sure what would be a better solution. I'm studying for my summer finals right now but i'll be back on the case by tuesday afternoon in theory.

**Comment by MikaelCluseau** [comment]
Hi @BenTheElder, I hope you finals went good :)

I still haven't found a better solution than either using userspace proxying or making the pod connect to himself by mapping the expected name to localhost in /etc/hosts (defeating the purpose of the "self connection check", if I understand it correctly; I may ask people of Apache Kafka about that, but it's another story).

Also, I think the interpretation I read somewhere on the Internets about "it's because a conntrack has been set" is wrong. I used a `TRACE` rule in mangle's `PREROUTING` table, and the life of the packet doesn't end after the `DNAT` rule in the normal case. The packet stops appearing in iptables after it has been DNATed specifically to the same IP than the source one. I reformated the trace logs below, using "___" when the value hasn't changed.

pod2-to-pod1 (the normal case):
```
Position                                |      IN |     OUT |      PHYSIN |                                       MAC |           SRC |            DST | LEN |  TOS | PREC | TTL |    ID | DF | PROTO |   SPT |  DPT |        SEQ | ACK | WINDOW |  RES | SYN | URGP | OPT | (020405820402080A19FA54340000000001030307) |     PHYSOUT
--------------------------------------- | ------- | ------- | ----------- | ----------------------------------------- | ------------- | -------------- | --- | ---- | ---- | --- | ----- | -- | ----- | ----- | ---- | ---------- | --- | ------ | ---- | --- | ---- | --- | ------------------------------------------ | -----------
raw:PREROUTING:policy:3                 | docker0 |         | vethb63f6b7 | 02:47:40:98:8f:97:02:42:0a:f4:25:da:08:00 | 10.244.37.218 | 10.107.241.117 |  60 | 0x00 | 0x00 |  64 | 30198 |    |   TCP | 34925 | 9092 | 2353471938 |   0 |  28200 | 0x00 |     |    0 |     |                                            |            
mangle:PREROUTING:policy:1              | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
nat:PREROUTING:rule:1                   | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
nat:KUBE-PORTALS-CONTAINER:rule:197     | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
nat:KUBE-SVC-9xs1gp9Yi0xCMiVqbUL:rule:1 | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
nat:KUBE-SVC-iWFvW7fh34r6zYRIvv0:rule:2 | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
mangle:FORWARD:policy:1                 | _______ | docker0 | ___________ | 02:42:0a:f4:25:7d:02:42:0a:f4:25:da:08:00 | _____________ |  10.244.37.125 | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | veth0fc21f5
filter:FORWARD:rule:1                   | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
filter:DOCKER:return:2                  | _______ | _______ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | __________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________ | ___________
```

pod1-to-pod1 (the problematic case):
```
Position                                |      IN | OUT |      PHYSIN |                                       MAC |           SRC |            DST | LEN |  TOS | PREC | TTL |    ID | DF | PROTO |   SPT |  DPT |       SEQ | ACK | WINDOW |  RES | SYN | URGP | OPT | (020405820402080A19F404C60000000001030307)
--------------------------------------- | ------- | --- | ----------- | ----------------------------------------- | ------------- | -------------- | --- | ---- | ---- | --- | ----- | -- | ----- | ----- | ---- | --------- | --- | ------ | ---- | --- | ---- | --- | ------------------------------------------
raw:PREROUTING:policy:2                 | docker0 |     | veth0fc21f5 | 02:47:40:98:8f:97:02:42:0a:f4:25:7d:08:00 | 10.244.37.125 | 10.107.241.117 |  60 | 0x00 | 0x00 |  64 | 14813 |    |   TCP | 58587 | 9092 | 414465307 |   0 |  28200 | 0x00 |     |    0 |     |                                           
mangle:PREROUTING:policy:1              | _______ | ___ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | _________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________
nat:PREROUTING:rule:1                   | _______ | ___ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | _________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________
nat:KUBE-PORTALS-CONTAINER:rule:197     | _______ | ___ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | _________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________
nat:KUBE-SVC-9xs1gp9Yi0xCMiVqbUL:rule:1 | _______ | ___ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | _________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________
nat:KUBE-SVC-iWFvW7fh34r6zYRIvv0:rule:1 | _______ | ___ | ___________ | _________________________________________ | _____________ | ______________ | ___ | ____ | ____ | ___ | _____ | __ | _____ | _____ | ____ | _________ | ___ | ______ | ____ | ___ | ____ | ___ | __________________________________________
```

* `nat:KUBE-SVC-9xs1gp9Yi0xCMiVqbUL:rule:1` is `-A KUBE-SVC-9xs1gp9Yi0xCMiVqbUL -m comment --comment "master/kafka1:" -j KUBE-SVC-iWFvW7fh34r6zYRIvv0`
* `nat:KUBE-SVC-iWFvW7fh34r6zYRIvv0:rule:1` is `-A KUBE-SVC-iWFvW7fh34r6zYRIvv0 -p tcp -m comment --comment "master/kafka1:" -m recent --set --name KUBE-SVC-iWFvW7fh34r6zYRIvv0 --mask 255.255.255.255 --rsource -j DNAT --to-destination 10.244.37.125:9092`

Hope this helps.

**Comment by MikaelCluseau** [comment]
A solution could be to manage the pod's iptable, if it's possible, and use a `REDIRECT` rule there.

If it's worth it, we could even handle the LB case this way, if we set up rules like this (for 3 hosts LB):
* in the pod, a probability rule, like `-j REDIRECT --probability 0.333`
* on the host, use a subset of the normal rules in this case, starting a probability 0.5, to the other replicas (so `-j DNAT --probability 0.5 --to-destination replica2:port` plus `-j DNAT --to-destination replica3:port`
This will require a specific set of rule per endpoint when the service has more than 1 endpoint, though.

**Comment by BenTheElder** [comment]
@MikaelCluseau thanks for updating me!
They went well (I think).

Is this still with flannel?

Hmm; With those rules, that seems to kind of break the expected round robin aspect, particularly for other services trying to access something that happens to be running on the same host. Also How would we handle single endpoint services?

**Comment by MikaelCluseau** [comment]
This is still with flannel, but I don't see where it could interact, especially when the packet is already traversing the rules. The only thing it does in iptables is:
```
-A POSTROUTING -s 10.244.0.0/16 -j FLANNEL
-A FLANNEL -d 10.244.0.0/16 -j ACCEPT
-A FLANNEL ! -d 224.0.0.0/4 -j MASQUERADE
```
The -j FLANNEL is the last rule in `POSTROUTING`.

I meant those rules only when the packet comes from one the endpoints. To be more verbose, what I'm thinking about, outside the pod, is a service rule (`KUBE-SVC-{{service}}`) like:
```
-s {replica1-ip} -j KUBE-SVC-{to-replica2} --probability 0.5
-s {replica1-ip} -j KUBE-SVC-{to-replica3} --probability 1

-s {replica2-ip} -j KUBE-SVC-{to-replica1} --probability 0.5
-s {replica2-ip} -j KUBE-SVC-{to-replica3} --probability 1

-s {replica3-ip} -j KUBE-SVC-{to-replica1} --probability 0.5
-s {replica3-ip} -j KUBE-SVC-{to-replica2} --probability 1

-j KUBE-SVC-{to-replica1} --probability 0.33
-j KUBE-SVC-{to-replica2} --probability 0.5
-j KUBE-SVC-{to-replica3} --probability 1
```

Which means that a packet to this service from replica would traverse this sequence:
1. on the pod: `-j REDIRECT --to... --probability 0.33`
2. on the host: `-s {replica1-ip} -j KUBE-SVC-{to-replica2} --probability 0.5`
3. on the host: `-s {replica1-ip} -j KUBE-SVC-{to-replica3} --probability 1`

The problem here is the O(n**2) rule complexity (a 100 replicas will produce 100 times 99 rules set for the specific case plus 1 time the 100 rule set for the normal case). But well, in this case, you have a service that requires a 100 replicas so it's probably not that a big deal. Also, we can lower the rule count by filtering the specific rule sets to the only endpoint that are local to the host we manage.

**Comment by BenTheElder** [comment]
Hmm. I'm not sure how reliably we could handle filtering though as we'd need to know all of the host addresses, and otherwise it would be very costly.

I'm still interested in why it won't work from pod1->pod as well, it seems odd that it works pod1->pod2 and then stalls pod1->pod1.

**Comment by BenTheElder** [comment]
What is different between `raw:PREROUTING:policy:2` and `raw:PREROUTING:policy:3` ?

**Comment by MikaelCluseau** [comment]
> I'm not sure how reliably we could handle filtering though as we'd need to know all of the host addresses, and otherwise it would be very costly.

I'm not sure as I haven't worked a lot in kubernetes, but the set of pods to handle the the set of pod that (1) are endpoints of the service and (2) have Status.HostIP = to our host IP (as given by the kubelet).

> I'm still interested in why

I'm too, but I don't have this answer (it's somewhere in Linux's sources of course).

> What is different between raw:PREROUTING:policy:2 and raw:PREROUTING:policy:3 ? 

One to match pod1-to-pod1 and one to match pod2-to-pod1 (tracing everything is not possible :)).

**Comment by BenTheElder** [comment]
Hmm. I'm going to see if I can ping @thockin on this one, this is a bit beyond my limited iptables knowledge, in the meantime I want to keep digging around to see if I can find anything but if not then I suppose we may have to fall back on something like the rule filtering you suggested. It'd be a shame to increase the complexity so much though.

**Comment by MikaelCluseau** [comment]
Yes, this why we need to see if it is worth the complexity. IMHO it's not and we should just add a rule in the container so it loops back to itself when it wants to use a service it provides, but I don't have the broad view to assert it's true in every use case.


Envoyé depuis un mobile.

-------- Message d'origine --------
De : Benjamin Elder <notifications@github.com> 
Date : 29/07/2015  11:24  (GMT+11:00) 
À : BenTheElder/kubernetes <kubernetes@noreply.github.com> 
Cc : MikaelCluseau <mcluseau@isi.nc> 
Objet : Re: [kubernetes] Pod-to-pod networking broken (#3) 

Hmm. I'm going to see if I can ping @thockin on this one, this is a bit beyond my limited iptables knowledge, in the meantime I want to keep digging around to see if I can find anything but if not then I suppose we may have to fall back on something like the rule filtering you suggested. It'd be a shame to increase the complexity so much though.

**Comment by BenTheElder** [comment]
I think that a rule to loop back to itself is pretty reasonable, I've still messaged thockin on freenode and started looking around some more to try and pin down a cause. I do wish we could get some more discussion going with other users, but I suppose they can just rewrite it how they want later if not.

**Comment by BenTheElder** [comment]
@MikaelCluseau have you seen @Symmetric's proposal? https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#issuecomment-125794747
Currently I've been planning to look into just adding the redirect rule and working on the iptables-save stuff but I wanted to get some more discussion going and the proposal there seems like an interesting possible change.

I also most certainly do not have the broad view to assert what we should be doing in all cases, perhaps if we can get more devs discussing the rules then we can assemble the broader view.

**Comment by Symmetric** [comment]
I _think_ these cases do not overlap (my suggestion is for nodePort, rather than hitting a service VIP from a pod in the service).

I've thought about this a bit, and I can't see a clean way to get round this problem. The trouble with NAT is that it's hard to tell a perfect lie...

Did you try something like `iptables -t nat -I POSTROUTING -s <pod IP> -d <pod IP> -j SNAT --to-source <svc IP>`? I suspect that might work for a single service, but it won't work in general since the pod could be referenced by multiple services, and you can't disambiguate which service the traffic was DNATed by. You might be able to get round this by MARKing packets before DNATing them, to specify which service's IP you need to SNAT in the src=dst=podIP case.

Anything that causes O(n^2) complexity in the number of iptables rules is very problematic -- n could be hundreds or conceivably thousands if you're running a big DC.

**Comment by MikaelCluseau** [comment]
Hi there, sorry to be late, I had issues here as the DC hosting us had multiple powerlosses...

> I think these cases do not overlap

They do not, although they can both require userland proxying unless a better iptables solution is found.

> Did you try something like iptables -t nat -I POSTROUTING -s <pod IP> -d <pod IP> -j SNAT --to-source <svc IP>?

Yes, that did not work because rule traversal stop after the DNAT did set the same source and destination IPs on the packet. Which means this...

> You might be able to get round this by MARKing packets before DNATing them

... did not work either (it was my second guess too ;)), because the rules are not traversed anymore (not even filter's FORWARD table), as can be seen in the trace tables I posted [earlier](https://github.com/BenTheElder/kubernetes/issues/3#issuecomment-125779469).

> Anything that causes O(n^2) complexity in the number of iptables rules is very problematic

For sure, especially when it's for a corner case.

> n could be hundreds or conceivably thousands if you're running a big DC

I'm not sure, here `n` is the number of replicas for the service; if you have `m` services with an average of `n` replicas, the rule count will be O(m*n**2) instead of O(m*n). And since a host do not have to create rules for pods running outside of him, n will 1 if the replicas are perfectly spread over the cluster (at most 1 pod per service per host). We will write twice as much rules, but this stays O(m*n). Even will userspace proxying, we will end up writing twice as much rules (1 `REDIRECT` rule per host pod). The only way to stay closer to the current rule count is to put a `REDIRECT` in the source pod than will unconditionnaly loop the service request back to him, which I think is completely acceptable: the case I hit is a service checking its "advertised hostname", and the more think about that case, the more I feel like its a mistake anyway (I will have to talk to Kafka people about that).

**Comment by BenTheElder** [comment]
@MikaelCluseau, noting that I'm following all of this discussion and have been working on getting/parsing and properly handling the results of `iptables-save` when syncing rules, but this is next on my todo list after that. I hit a snag with increased memory requirements bugging the build when I rebased against master again earlier but I'm back on track now.

I currently think the redirect in source pod seems pretty reasonable, if it's trying to connect to itself anyhow then we can just redirect to it, but I'm not sure about the other rule changes or alternate solutions.
  
  
I'm hoping to have those two checked off soon (by Monday perhaps?), and once I at least have the iptables-save stuff complete I'll ping thockin again and see if I can get the review started. If we decide we need more rule changes that should help get more discussion going on which tradeoffs to make etc...

**Comment by MikaelCluseau** [comment]
@BenTheElder no worries, I solved the specific problem here by adding a /etc/hosts entry (side note: not having any userspace proxy dropped our average "full loop" response time from 4-5s to &lt;100ms, so its pretty huge). So, monday or anytime later is ok to me. I just want to make clear that I'm not putting any pressure here. In fact, I'm even pretty happy that it broke the Kafka service because I was wondering why it was still using the (userland) kube-proxy before. I'm putting myself in standby mode until you have the redirect rule for me test, but I can't promise I won't try some things in the mean time, and report here :-)

**Comment by BenTheElder** [comment]
Thanks.
That sounds like an impressive improvement!
I don't feel under any pressure, but I would like to get the PR tidied up
for review before GSoC ends.

**Comment by thockin** [comment]
I'm finally caught up a bit, but this is a long issue - can someone summarize for me?

**Comment by BenTheElder** [comment]
In @MikaelCluseau's testing, a pod connecting back to itself will have the packet hang.
[This comment](https://github.com/BenTheElder/kubernetes/issues/3#issuecomment-125779469) has trace information from doing `pod2 -> pod1` and `pod1 -> pod1`.

**Comment by BenTheElder** [comment]
In further summary, he's currently falling back on a host entry and the plan to try and solve it is to handle `pod->self` with a redirect rule, but I don't really know what the root of the problem is yet so that is tentative. 

I'm in the process of moving atm but Sunday I'll start trying to debug this further after I finish up making use of the iptables-save state to preserve counters etc.

**Comment by thockin** [comment]
regarding the syscall: You could have the kube-proxy exec() modprobe for
the module, ignore failures, then blast the value into the that sysctl if
it is not correct.  And log that you did that.

I'm trying to repro the self-endpoint problem.  I see it (strangely) hang
for a second then try another endpoint, so it work but with some latency.
Will keep investigating.

**Comment by BenTheElder** [comment]
Thanks for looking into this! I'll put the sysctl stuff on my todo list for
this week.

Have not reproduced the other issue yet as my local branch has been in a
dirty state, but if you or anyone else doesn't get to it before my next
commit I'll start trying to reproduce and debug well.

**Comment by thockin** [comment]
The retry I observed seems to be inside wget.  Sigh.  I can repro easily with a single specific rule to cause it.  I have not found an answer yet.

Anything involving putting iptables rules into pods is sketchy at best - a pod can be in multiple services at once, and that can change dynamically.  We really don't want to be in the business of switching namespaces underneath docker if we can avoid it. 

As for the rest of this work - I apologize for largely ignoring it most of the summer - 1.0 launch was nuts.  It is my intent to give this as much attaention as I can henceforth until it is merged or at least I could take it over to iron out last wrinkles.  I'm deferring all other kube-proxy work until then.  Have time to chat this week?

**Comment by thockin** [comment]
Aside: the note aboutthe new kernels requiring a modprobe will (I think) fix an entirely unrelated problem I am having!  Thanks for the clue!!  We'll need to load that modula and set that sysctl for kubernetes in general.

**Comment by BenTheElder** [comment]
Ah, too bad. I'm not really sure what to do then, my iptables knowledge is
pretty limited in comparison to the three of you.

And understandable, If I sounded antsy it's because I wanted to make sure
things were headed in the right direction, but I could see how swamped k8s
has been in pull requests, issues, and the launch etc.

I definitely have time to chat this week.

**Comment by thockin** [comment]
A solution.  We have to enable hairpin mode on bridge interfaces and we
have to masquerade access-to-self.  Observe:

```
thockin@kubernetes-minion-6b59:~$ sudo bash

root@kubernetes-minion-6b59:/home/thockin# iptables-save | grep 9.9.9.9

root@kubernetes-minion-6b59:/home/thockin# ID=$(docker run -d
thockin/serve_hostname:thtest)
Unable to find image 'thockin/serve_hostname:thtest' locally
thtest: Pulling from thockin/serve_hostname
cf2616975b4a: Already exists
6ce2e90b0bc7: Already exists
8c2e06607696: Already exists
109f47b829e0: Already exists
26fa00b034bd: Already exists
452f40184a2a: Already exists
748b0d776d7f: Already exists
4aab7f9c61b7: Already exists
4aab7f9c61b7: Already exists
Digest:
sha256:d7e0d8ac048db344981ec2005b8d9e1e72868fa1a1e2ba513e45811b942b235a
Status: Image is up to date for thockin/serve_hostname:thtest

root@kubernetes-minion-6b59:/home/thockin# echo $ID
df4cc3092b3817ec202259f06464e5b87d6b8b4db3654143662b56fb9d71c154

root@kubernetes-minion-6b59:/home/thockin# IP=$(docker exec $ID hostname -i)

root@kubernetes-minion-6b59:/home/thockin# echo $IP
10.244.3.8

root@kubernetes-minion-6b59:/home/thockin# iptables -t nat -A PREROUTING -d
9.9.9.9/32 -p tcp -m state --state NEW -m tcp --dport 80 -j DNAT
--to-destination $IP:9376

root@kubernetes-minion-6b59:/home/thockin# iptables -t nat -A OUTPUT -d
9.9.9.9/32 -p tcp -m state --state NEW -m tcp --dport 80 -j DNAT
--to-destination $IP:9376

root@kubernetes-minion-6b59:/home/thockin# curl 9.9.9.9; echo
df4cc3092b38

root@kubernetes-minion-6b59:/home/thockin# docker exec $ID wget -qO-
9.9.9.9; echo
^C

root@kubernetes-minion-6b59:/home/thockin# iptables -t nat -A POSTROUTING
-s $IP/32 -d $IP/32 -p tcp -m tcp --dport 9376 -j MASQUERADE

root@kubernetes-minion-6b59:/home/thockin# docker exec $ID wget -qO-
9.9.9.9; echo
^C

root@kubernetes-minion-6b59:/home/thockin# brctl showstp cbr0 | grep -i
hairpin

root@kubernetes-minion-6b59:/home/thockin# for intf in $(ip link list |
grep veth | cut -f2 -d:); do brctl hairpin cbr0 $intf on; done

root@kubernetes-minion-6b59:/home/thockin# brctl showstp cbr0 | grep -i
hairpin
 hairpin mode   1
 hairpin mode   1
 hairpin mode   1
 hairpin mode   1
 hairpin mode   1

root@kubernetes-minion-6b59:/home/thockin# docker exec $ID wget -qO-
9.9.9.9; echo
^C

root@kubernetes-minion-6b59:/home/thockin# cat
/proc/sys/net/ipv4/conf/cbr0/route_localnet
0

root@kubernetes-minion-6b59:/home/thockin# echo 1 >
/proc/sys/net/ipv4/conf/cbr0/route_localnet

root@kubernetes-minion-6b59:/home/thockin# cat
/proc/sys/net/ipv4/conf/cbr0/route_localnet
1

root@kubernetes-minion-6b59:/home/thockin# docker exec $ID wget -qO-
9.9.9.9; echo
df4cc3092b38
```

Can you guys repro to confirm, please?

Now how to actually deliver this?  Adding the masquerade on access-to-self
is easy and can be part of this patch set.  Adding the brctl change is a
little trickier.  We need to do that for every interface, which really
means it needs to be a network plugin or something.  I think we could just
add it in the network plugin path.

Assuming you can confirm this finding, we can work it into this patch set,
too.

**Comment by Symmetric** [comment]
@thockin is it an issue that you needed to use route_localnet (since that's not included in Red Hat's 2.6 kernel)?

I'll need to test this fix with Calico (which doesn't use the bridge); I think that the masquerade rule alone should work in our case since we don't have the bridge in the path.

**Comment by MikaelCluseau** [comment]
Hi @thockin, I can confirm it solves the problem, the packet does the expected traversal just after I enabled the hairpin mode (I tried setting route_localnet before enabling hairpin to be sure hairpin was required). So, the POSTROUTING show this:

```
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

    1    60 MASQUERADE  all  --  *      *       10.244.37.128        10.244.37.128
```

and the self-connection works as expected:

```
$ kubectl --namespace=master exec keystone-yukci -- curl --connect-timeout 1 -Is keystone:35357
HTTP/1.1 300 Multiple Choices
Vary: X-Auth-Token
X-Distribution: Ubuntu
Content-Type: application/json
Content-Length: 753
Date: Mon, 03 Aug 2015 20:13:17 GMT
```

That's great, thanks :-)

**Comment by MikaelCluseau** [comment]
FWIW, I found it easy to understand what hairpin does and how it relates to this issue as I found this:

https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/net/bridge/br_forward.c#n30

**Comment by BenTheElder** [comment]
So to recap:
- For the nf-bridge modprobe and sysctl we will add this somewhere kubernetes wise to fix the first issue.
- For the route to self, we will enable hairpin mode on bridges and set:  
`-A POSTROUTING -s <hostIP>/32 -d <hostIP>/32 -p tcp -m tcp --dport 9376 -j MASQUERADE` on the host?
Do we also need this for udp?

**Comment by thockin** [comment]
I think that kube v1.1 will either not support CentOS or will have a
fallback mode for kube-proxy.  We need to figure out a position on this.

**Comment by BenTheElder** [comment]
I currently select the new proxy or the old proxy depending on the iptables
version (this needs discussing), so right now it should fall back on the
userspace proxy.

**Comment by thockin** [comment]
Another approach would be to use the pure proxy and install a userspace
proxy JUST for that one case.

**Comment by MikaelCluseau** [comment]
Won't the iptables rules for the userspace mode conflict with pure-iptables version rules?

Also, if I can avoid having the userspace proxy for just one case that doesn't affect me as a CoreOS user, it would be better (though not critical of course).

**Comment by BenTheElder** [comment]
I've been trying to handle the `MASQUERADE` rule but running into [this](https://github.com/GoogleCloudPlatform/kubernetes/issues/12285) has made spinning up a cluster impossible for the moment. So unfortunately while I think I've had it working, I still can't verify it (yet).

@thockin, otherwise though, the PR should be essentially ready for review, since you've said we will handle the hairpin and nf-bridge stuff elsewhere.

**Comment by BenTheElder** [comment]
@thockin @MikaelCluseau Do we / should we have upstream issues for the hairpin mode and nf-bridge?

We should probably get those taken care of along with the PR; I plan to go ahead add a masquerade rule to the new proxy's `iptablesInit` along the lines of:
`-A POSTROUTING -s <hostIP>/32 -d <hostIP>/32 -j MASQUERADE`
So that we can have this taken care of when the PR goes in.

**Comment by BenTheElder** [comment]
I think we can safely say this is being taken care of upstream. Closing this here.
